corehq/apps/hqcase/api/get_list.py

```diff
@@ -0,0 +1,96 @@
+import datetime
+from itertools import chain
+
+from dateutil.parser import parse
+
+from dimagi.utils.parsing import FALSE_STRINGS
+
+from corehq.apps.es import case_search
+from corehq.apps.es import cases as case_es
+
+from .core import UserError, serialize_es_case
+
+DEFAULT_PAGE_SIZE = 20
+MAX_PAGE_SIZE = 5000
+CUSTOM_PROPERTY_PREFIX = 'property.'
+
+
+def _to_boolean(val):
+    return not (val == '' or val.lower() in FALSE_STRINGS)
+
+
+def _to_int(val, param_name):
+    try:
+        return int(val)
+    except ValueError:
+        raise UserError(f"'{val}' is not a valid value for '{param_name}'")
+
+
+def _make_date_filter(date_filter, param):
+
+    def filter_fn(val):
+        try:
+            # If it's only a date, don't turn it into a datetime
+            val = datetime.datetime.strptime(val, '%Y-%m-%d').date()
+        except ValueError:
+            try:
+                val = parse(val)
+            except ValueError:
+                raise UserError(f"Cannot parse datetime '{val}'")
+        return date_filter(**{param: val})
+
+    return filter_fn
+
+
+def _to_date_filters(field, date_filter):
+    return [
+        (f'{field}.gt', _make_date_filter(date_filter, 'gt')),
+        (f'{field}.gte', _make_date_filter(date_filter, 'gte')),
+        (f'{field}.lte', _make_date_filter(date_filter, 'lte')),
+        (f'{field}.lt', _make_date_filter(date_filter, 'lt')),
+    ]
+
+
+FILTERS = {
+    'external_id': case_search.external_id,
+    'case_type': case_es.case_type,
+    'owner_id': case_es.owner,
+    'case_name': case_es.case_name,
+    'closed': lambda val: case_es.is_closed(_to_boolean(val)),
+}
+FILTERS.update(chain(*[
+    _to_date_filters('last_modified', case_es.modified_range),
+    _to_date_filters('server_last_modified', case_es.server_modified_range),
+    _to_date_filters('date_opened', case_es.opened_range),
+    _to_date_filters('date_closed', case_es.closed_range),
+]))
+
+
+def get_list(domain, params):
+    start = _to_int(params.pop('offset', 0), 'offset')
+    page_size = _to_int(params.pop('limit', DEFAULT_PAGE_SIZE), 'limit')
+    if page_size > MAX_PAGE_SIZE:
+        raise UserError(f"You cannot request more than {MAX_PAGE_SIZE} cases per request.")
+
+    query = (case_search.CaseSearchES()
+             .domain(domain)
+             .size(page_size)
+             .start(start)
+             .sort("@indexed_on"))
+
+    for key, val in params.items():
+        if key.startswith(CUSTOM_PROPERTY_PREFIX):
```

Comment:
YAGNI, maybe, but it would be really fun if you also allowed an "xpath query" like the CLE does. I think would just be 2 lines of code here:
```python
if key == 'xpath':
    query = query.xpath_query(domain, val)
```
And would give you all the power of the CLE through the API.
```
case/?xpath="(foo_date >= '2020-01-01' and bar_date <= '2020-12-01' or name='harmon')"
```

Reply:
Ha, I was thinking that attempting to put an XPath expression in a URL querystring would be foolhardy, but then I realized that the CLE already does that!  Perfect, I'll add that in once I [settle on a pattern](https://github.com/dimagi/commcare-hq/pull/29275#discussion_r587285811)

Reply:
Another plus side of this would be that it'd allow for more sophisticated case property filtering than exact matches.  Support for that would be great, but I also didn't want to invent a new one-off grammar for that sort of thing that lives only in URL querystrings.

Reply:
I don't think we allow non-exact matches through this currently. We were thinking of adding something like `~=` to the xpath parser to imply a fuzzy match, but haven't found it to useful to anyone yet.

Reply:
added!